### PR TITLE
lib.systems.doubles.all: reorganize

### PR DESCRIPTION
**Pull request by alyssais**
###### Motivation for this change

The previous mess was partially grouped by OS, and partially grouped
by architecture, which made it very difficult to know where to add new
entries.

I've chosen to group by OS entirely, because OSes are likely to
maintain exhaustive lists of supported architectures, but it's far
less likely we'd be able to find exhaustive lists of supported OSes
for every architecture.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: lib/systems/doubles.nix [hunk 1]
```diff
@@ -3,53 +3,53 @@ let
...
 
-    "wasm64-wasi" "wasm32-wasi"
+    linux = [
+      "aarch64-linux" "armv5tel-linux" "armv6l-linux" "armv7a-linux"
+      "armv7l-linux" "i686-linux" "mipsel-linux" "powerpc64-linux"
+      "powerpc64le-linux" "riscv32-linux" "riscv64-linux" "x86_64-linux"
+    ];
 
-    "x86_64-redox"
+    mmix = [ "mmix-mmixware" ];
```
Comment (OPNA2608) [inline lib/systems/doubles.nix:27]: I think MMIX is the architecture's name, not the OS?

Comment (alyssais) [reply lib/systems/doubles.nix]: You're absolutely right.  Fixed, and thanks for catching it.

**Comment by Ericson2314** [comment]
I'm fine with rearranging the big lists, but I think all lists below should be calculated not handwritten. This is important to ensure that the predicates and the platforms list stay in sync.

**Comment by Ericson2314** [comment]
See also https://github.com/NixOS/nixpkgs/blob/master/lib/tests/systems.nix for the lists which are supposed to stay hand coded.

**Comment by alyssais** [comment]
> I'm fine with rearranging the big lists, but I think all lists below should be calculated not handwritten. This is important to ensure that the predicates and the platforms list stay in sync.

I don't understand what you mean by this.  Can you rephrase?

**Comment by Ericson2314** [comment]
I mean when we do e.g.
```
darwin = filterDoubles predicates.isDarwin;
```
and then in https://github.com/NixOS/nixpkgs/blob/master/lib/tests/systems.nix do
```
testdarwin = mseteq darwin [ "x86_64-darwin" "i686-darwin" "aarch64-darwin" "armv7a-darwin" ];
```
we are testing `isDarwin` does what we expect, but if we write
```
darwin = [ "x86_64-darwin" "i686-darwin" "aarch64-darwin" "armv7a-darwin" ];
```
here we are no longer testing `isDarwin`.

**Comment by alyssais** [comment]
> I mean when we do e.g.
> ```
> darwin = filterDoubles predicates.isDarwin;
> ```
> and then in https://github.com/NixOS/nixpkgs/blob/master/lib/tests/systems.nix do
> ```
> testdarwin = mseteq darwin [ "x86_64-darwin" "i686-darwin" "aarch64-darwin" "armv7a-darwin" ];
> ```
> we are testing `isDarwin` does what we expect, but if we write
> ```
> darwin = [ "x86_64-darwin" "i686-darwin" "aarch64-darwin" "armv7a-darwin" ];
> ```
> here we are no longer testing `isDarwin`.

Ah, I see.  So should I just make `all` a big array, with comments to
seperate different OSes maybe, and go back to generating all the smaller
lists?

**Comment by Ericson2314** [comment]
That sounds great to me!
